proxy-socket-core/src/main/java/net/airvantage/proxysocket/core/v2/ProxyProtocolV2Decoder.java

```diff
@@ -0,0 +1,231 @@
+        final int bytesConsumed;
+
+        AddressPair(InetSocketAddress src, InetSocketAddress dst, int bytesConsumed) {
+            this.src = src;
+            this.dst = dst;
+            this.bytesConsumed = bytesConsumed;
+        }
+    }
+
+    private static AddressPair parseAddresses(byte[] data, int currentPosition, ProxyHeader.AddressFamily af, int variableLength)
+            throws ProxyProtocolParseException {
+        return switch (af) {
+            case AF_INET -> parseIPv4Addresses(data, currentPosition, variableLength);
+            case AF_INET6 -> parseIPv6Addresses(data, currentPosition, variableLength);
+            case AF_UNIX -> parseUnixAddresses(data, currentPosition, variableLength);
+            default -> throw new ProxyProtocolParseException("Invalid address family");
+        };
+    }
+
+    private static int IPV4_ADDR_PAIR_LEN = 2*(IPV4_ADDR_LEN + PORT_LEN);
+    private static AddressPair parseIPv4Addresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < IPV4_ADDR_PAIR_LEN) {
+            throw new ProxyProtocolParseException("Truncated IPv4 address block in header");
+        }
+
+        InetAddress s;
+        InetAddress d;
+        try {
+            s = InetAddress.getByAddress(new byte[]{data[pos++], data[pos++], data[pos++], data[pos++]});
+            d = InetAddress.getByAddress(new byte[]{data[pos++], data[pos++], data[pos++], data[pos++]});
+        } catch (UnknownHostException e) {
+            throw new ProxyProtocolParseException("Invalid IPv4 address in header", e);
+        }
+
+        int sp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+        int dp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+
+        return new AddressPair(new InetSocketAddress(s, sp), new InetSocketAddress(d, dp), IPV4_ADDR_PAIR_LEN);
+    }
+
+    private static int IPV6_ADDR_PAIR_LEN = 2*(IPV6_ADDR_LEN + PORT_LEN);
+    private static AddressPair parseIPv6Addresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < IPV6_ADDR_PAIR_LEN) {
+            throw new ProxyProtocolParseException("Truncated IPv6 address block in header");
+        }
+
+        InetAddress s;
+        InetAddress d;
+        byte[] sb = new byte[IPV6_ADDR_LEN];
+        byte[] db = new byte[IPV6_ADDR_LEN];
+        System.arraycopy(data, pos, sb, 0, IPV6_ADDR_LEN);
+        System.arraycopy(data, pos+IPV6_ADDR_LEN, db, 0, IPV6_ADDR_LEN);
+        try {
+            s = InetAddress.getByAddress(sb);
+            d = InetAddress.getByAddress(db);
+        } catch (UnknownHostException e) {
+            throw new ProxyProtocolParseException("Invalid IPv6 address in header", e);
+        }
+
+        pos += 2*IPV6_ADDR_LEN;
+        int sp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+        int dp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
```

Comment:
I think the last ++ of poss++ is useless but maybe we can keep for readability ?